store/tooling/tooling.go

```diff
@@ -0,0 +1,475 @@
+		return nil, nil
+	}
+	if opts.BeforeDownloadFunc == nil {
+		return nil, fmt.Errorf("internal error: DownloadManyOptions.BeforeDownloadFunc must be set")
+	}
+
+	actionFlag := store.SnapActionIgnoreValidation
+	if opts.EnforceValidation {
+		actionFlag = store.SnapActionEnforceValidation
+	}
+
+	downloadedSnaps = make(map[string]*DownloadedSnap, len(toDownload))
+	current := make([]*store.CurrentSnap, 0, len(curSnaps))
+	for _, csnap := range curSnaps {
+		ch := "stable"
+		if csnap.Channel != "" {
+			ch = csnap.Channel
+		}
+		current = append(current, &store.CurrentSnap{
+			InstanceName:     csnap.SnapName,
+			SnapID:           csnap.SnapID,
+			Revision:         csnap.Revision,
+			TrackingChannel:  ch,
+			Epoch:            csnap.Epoch,
+			IgnoreValidation: !opts.EnforceValidation,
+		})
+	}
+
+	actions := make([]*store.SnapAction, 0, len(toDownload))
+	for _, sn := range toDownload {
+		actions = append(actions, &store.SnapAction{
+			Action:       "download",
+			InstanceName: sn.Snap.SnapName(), // XXX consider using snap-id first
+			Channel:      sn.Channel,
+			CohortKey:    sn.CohortKey,
+			Flags:        actionFlag,
+		})
+	}
+
+	sars, _, err := tsto.sto.SnapAction(context.TODO(), current, actions, nil, tsto.user, nil)
+	if err != nil {
+		// err will be 'cannot download snap "foo": <reasons>'
+		return nil, err
+	}
+
+	for _, sar := range sars {
+		targetPath, err := opts.BeforeDownloadFunc(sar.Info)
+		if err != nil {
+			return nil, err
+		}
+		dlSnap, err := tsto.snapDownload(targetPath, &sar, DownloadSnapOptions{})
+		if err != nil {
+			return nil, err
+		}
+		downloadedSnaps[sar.SnapName()] = dlSnap
+	}
+
+	return downloadedSnaps, nil
+}
+
+// AssertionFetcher creates an asserts.Fetcher for assertions against the given store using dlOpts for authorization, the fetcher will add assertions in the given database and after that also call save for each of them.
+func (tsto *ToolingStore) AssertionFetcher(db *asserts.Database, save func(asserts.Assertion) error) asserts.Fetcher {
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		return tsto.sto.Assertion(ref.Type, ref.PrimaryKey, tsto.user)
+	}
+	save2 := func(a asserts.Assertion) error {
+		// for checking
+		err := db.Add(a)
+		if err != nil {
+			if _, ok := err.(*asserts.RevisionError); ok {
+				return nil
+			}
+			return fmt.Errorf("cannot add assertion %v: %v", a.Ref(), err)
+		}
+		return save(a)
+	}
+	return asserts.NewFetcher(db, retrieve, save2)
+}
+
+// Find provides the snapsserts.Finder interface for snapasserts.DerviceSideInfo
+func (tsto *ToolingStore) Find(at *asserts.AssertionType, headers map[string]string) (asserts.Assertion, error) {
+	pk, err := asserts.PrimaryKeyFromHeaders(at, headers)
+	if err != nil {
+		return nil, err
+	}
+	return tsto.sto.Assertion(at, pk, tsto.user)
+}
+
+func MockToolingStore(sto Store) *ToolingStore {
+	osutil.MustBeTestBinary("tooling store can only be mocked in tests")
+	return &ToolingStore{sto: sto}
+}
```

Comment:
Note, this is a new addition to allow mocking from image package, it was previously only part for export test package.